src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/RawBinaryReader.cpp

```diff
@@ -0,0 +1,245 @@
+ * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The code contained herein was partially funded by the following contracts:
+ *    United States Air Force Prime Contract FA8650-07-D-5800
+ *    United States Air Force Prime Contract FA8650-10-D-5210
+ *    United States Prime Contract Navy N00173-07-C-2068
+ *    United States Air Force Prime Contract FA8650-15-D-5231
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#include "RawBinaryReader.hpp"
+
+#include <algorithm>
+#include <cstddef>
+
+#include "complex/Common/ComplexConstants.hpp"
+#include "complex/Common/Types.hpp"
+#include "complex/DataStructure/DataArray.hpp"
+#include "complex/DataStructure/DataStore.hpp"
+
+#if defined(_MSC_VER)
+const auto FSEEK = _fseeki64;
+#else
+const auto FSEEK = std::fseek;
+#endif
+
+namespace complex
+{
+
+namespace
+{
+#ifdef CMP_WORDS_BIGENDIAN
+constexpr int32_t k_EndianCheck = 0;
+#else
+constexpr int32_t k_EndianCheck = 1;
+#endif
+
+constexpr int32_t k_RbrNoError = 0;
+constexpr int32_t k_RbrFileNotOpen = -1000;
+constexpr int32_t k_RbrFileTooSmall = -1010;
+constexpr int32_t k_RbrFileTooBig = -1020;
+constexpr int32_t k_RbrDAError = -1040;
+constexpr int32_t k_RbrComponentError = -1050;
+constexpr int32_t k_RbrDANull = -1060;
+
+// -----------------------------------------------------------------------------
+int32_t SanityCheckFileSizeVersusAllocatedSize(size_t allocatedBytes, size_t fileSize, size_t skipHeaderBytes)
+{
+  if(fileSize - skipHeaderBytes < allocatedBytes)
+  {
+    return -1;
+  }
+  if(fileSize - skipHeaderBytes > allocatedBytes)
+  {
+    return 1;
+  }
+  // File Size and Allocated Size are equal so we are good to go
+  return 0;
+}
+
+// -----------------------------------------------------------------------------
+template <typename T>
+int32_t readBinaryFile(IDataArray* dataArrayPtr, const std::string& filename, uint64_t skipHeaderBytes, ChoicesParameter::ValueType endian)
+{
+  constexpr size_t k_DefaultBlocksize = 1048576;
+
+  DataArray<T>* dataArray = dynamic_cast<DataArray<T>*>(dataArrayPtr);
+
+  if(dataArray == nullptr)
+  {
+    return k_RbrDANull;
+  }
+
+  const size_t fileSize = std::filesystem::file_size(filename);
+  const size_t numBytesToRead = dataArray->getSize() * sizeof(T);
+  int32_t err = SanityCheckFileSizeVersusAllocatedSize(numBytesToRead, fileSize, skipHeaderBytes);
+
+  if(err < 0)
+  {
+    return k_RbrFileTooSmall;
+  }
+  if(err > 0)
+  {
+    return k_RbrFileTooBig;
+  }
+
+  FILE* f = std::fopen(filename.c_str(), "rb");
+  if(f == nullptr)
+  {
+    return k_RbrFileNotOpen;
+  }
+
+  // Skip some header bytes if the user asked for it.
+  if(skipHeaderBytes > 0)
+  {
+    FSEEK(f, skipHeaderBytes, SEEK_SET);
+  }
+
+  //  std::byte* chunkptr = reinterpret_cast<std::byte*>(dataArray->data());
+  std::byte* chunkptr = reinterpret_cast<std::byte*>(dataArray->template getIDataStoreAs<DataStore<T>>()->data());
```

Comment:
This isn't probably something we can address now, but this won't work with out of core DataStore types that we might add in the future.